libnetwork/iptables/firewalld.go

```diff
@@ -159,13 +197,44 @@ func OnReloaded(callback func()) {
 }
 
 // Call some remote method to see whether the service is actually running.
-func checkRunning() bool {
+func checkRunning() (bool, error) {
+
+	var currentState dbus.Variant
+	var wg sync.WaitGroup
+	var cancel context.CancelFunc
+	var ctx context.Context
+	var dbusTimeout = true
+	var err error
+
 	if connection == nil {
-		return false
+		return false, nil
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), DbusConnectionTimeout)
+	defer cancel()
+
+	//This is more complicated but without this, if polkit has an error dbus calls can hang for extended periods of time
+	wg.Add(1)
+
+	go func(ctx context.Context) {
+		defer wg.Done()
+		currentState, err = connection.sysObj.GetProperty(dbusInterface + ".state")
+		dbusTimeout = false
+	}(ctx)
+
+	wg.Wait()
+	if dbusTimeout {
+		return false, fmt.Errorf("Dbus Connection Timeout")
+	}
+
+	if err == nil {
+		value, ok := currentState.Value().(string)
+
+		// We are supposed to make sure firewalld is not in an INIT state
+		return ok && (value == "RUNNING" || value == "FAILED"), nil
```

Comment:
With my comment above; obtaining the value could be part of the go routine perhaps, so we'd have the value at the ready here (but have a closer look if that would make sense / would work)

Reply:
I wanted the goroutine to have a single purpose for clarity.  Ensuring the GetProperty call returning within the dbustimeout.